File: Doodle drifters 2/Assets/Scripts/State Machine/Client State Machine/StateMachine.js
import GameContext from "./GameContext.js";
import State from "./State.js";
import LobbyState from "./States/LobbyState.js"
import TestState from "./States/TestState.js"

/**
 * Enum replacement
 */
export const GameStateType = Object.freeze({
    LOBBY: "Lobby",
    DRAWING: "Drawing",
    VOTING: "Voting"
});

export default class StateMachine {
    constructor() {
        this.states = new Map();
        this.validTransitions = new Map([
            [GameStateType.LOBBY, [GameStateType.DRAWING]],
            [GameStateType.DRAWING, [GameStateType.VOTING]],
            [GameStateType.VOTING, [GameStateType.DRAWING]]
        ]);

        this.currentState = null;
        this.currentStateType = null;
        this.context = new GameContext(this);
        this.stateListeners = new Set();

        this.initializeStates();
    }

    initializeStates() {
        // Replace these with your real state implementations
        this.states.set(GameStateType.LOBBY, new LobbyState());
        this.states.set(GameStateType.DRAWING, new TestState());
        this.states.set(GameStateType.VOTING, new TestState());
    }

    start() {
        this.switchState(GameStateType.LOBBY);
    }

    getCurrentStateType() {
        return this.currentStateType;
    }

    update() {
        if (this.currentState) {
            this.currentState.update(this.context);
        }
    }

    fixedUpdate() {
        if (this.currentState) {
            this.currentState.fixedUpdate(this.context);
        }
    }

    switchState(newState) {
        if (this.currentState && newState === this.currentStateType) {
            return;
        }

        // Enable this when transitions are enforced
        /*
        const allowed = this.validTransitions.get(this.currentStateType);
        if (allowed && !allowed.includes(newState)) {
            console.warn("Invalid transition");
            return;
        }
        */

        const previousState = this.currentStateType;
        this.currentStateType = newState;
        this.setState(this.states.get(newState));
        this.emitStateChanged(previousState, this.currentStateType);
    }

    setState(newState) {
        if (this.currentState) {
            this.currentState.exit(this.context);
        }

        this.currentState = newState;

        if (this.currentState) {
            this.currentState.enter(this.context);
        }
    }

    onStateChanged(handler) {
        if (typeof handler !== "function") {
            return;
        }

        this.stateListeners.add(handler);
    }

    offStateChanged(handler) {
        this.stateListeners.delete(handler);
    }

    emitStateChanged(previousState, nextState) {
        this.stateListeners.forEach((handler) => {
            handler({ previousState, nextState, context: this.context });
        });
    }
}

/**
 * Example placeholder states
 * Replace these with your actual implementations
 */

File: Doodle drifters 2/Assets/Scripts/State Machine/Host State Machine/StateMachine.cs
using System.Collections.Generic;
using Unity.VisualScripting;
using UnityEngine;

public class StateMachine : MonoBehaviour
{
    public enum GameStateType
    {
        Lobby,
        NewScene,
        Think,
        Drawing,
        Voting,
        Solution
        //Add more types as neccesary
    }

    private Dictionary<GameStateType, IState> states;

    void InitializeStates()
    {
        states = new Dictionary<GameStateType, IState>
        { // change the actual states when they are created
            { GameStateType.Lobby, new LobbyState() },
            { GameStateType.NewScene, new NewSceneState() },
            { GameStateType.Think, new ThinkState() },
            { GameStateType.Drawing, new DrawingState() },
            { GameStateType.Voting, new VotingState() },
            { GameStateType.Solution, new SolutionState() },
        };
    }

    private Dictionary<GameStateType, List<GameStateType>> validTransitions =
        new Dictionary<GameStateType, List<GameStateType>>
    {
    { GameStateType.Lobby, new List<GameStateType> { GameStateType.NewScene } },
    { GameStateType.NewScene, new List<GameStateType> { GameStateType.Think } },
    { GameStateType.Think, new List<GameStateType> { GameStateType.Drawing } },
    { GameStateType.Drawing, new List<GameStateType> { GameStateType.Voting } },
    { GameStateType.Voting, new List<GameStateType> { GameStateType.Solution } },
    { GameStateType.Solution, new List<GameStateType> { GameStateType.Lobby } }
    };

    IState currentState;
    private GameContext context;
    private GameStateType currentStateType;

    void Awake()
    {
        context = new GameContext(this);
        InitializeStates();
    }

    void Start()
    {
        SwitchState(GameStateType.Lobby);
    }

    void Update()
    {
        currentState?.Update(context);
    }

    void FixedUpdate()
    {
        currentState?.FixedUpdate(context);
    }

    public void SwitchState(GameStateType newState)
    {
        // if (validTransitions[currentStateType].Contains(newState))
        // {
        //     Debug.LogWarning("invalid transition");
        //     return;
        // } uncomment when state transitions are fleshed out

        currentStateType = newState;
        SetState(states[newState]);
    }

    public void SetState(IState newState)
    {
        currentState?.Exit(context);
        currentState = newState;
        currentState?.Enter(context);
    }
}